operator/pkg/central/extensions/reconcile_central_db_password.go

```diff
@@ -10,57 +11,125 @@ import (
 	commonExtensions "github.com/stackrox/rox/operator/pkg/common/extensions"
 	"github.com/stackrox/rox/operator/pkg/types"
 	"github.com/stackrox/rox/pkg/renderer"
+	coreV1 "k8s.io/api/core/v1"
 	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
-	centralDBPasswordKey          = `password`
-	centralDBPasswordResourceName = "central-db-password"
+	centralDBPasswordKey = `password`
+
+	// canonicalCentralDBPasswordSecretName is the name of the secret that is mounted into Central (and Central DB).
+	// This is not configurable; if a user specifies a different password secret, the password from that needs to be
+	// mirrored into the canonical password secret.
+	canonicalCentralDBPasswordSecretName = `central-db-password`
 )
 
-// ReconcileCentralDBPasswordExtension returns an extension that takes care of creating the central-db-password
-// secret ahead of time.
+// ReconcileCentralDBPasswordExtension returns an extension that takes care of reconciling the central-db-password secret.
 func ReconcileCentralDBPasswordExtension(client ctrlClient.Client) extensions.ReconcileExtension {
 	return wrapExtension(reconcileCentralDBPassword, client)
 }
 
-func reconcileCentralDBPassword(ctx context.Context, c *platform.Central, client ctrlClient.Client, _ func(updateStatusFunc), _ logr.Logger) error {
-	if !c.Spec.Central.CentralDBEnabled() || c.Spec.Central.DB.IsExternal() {
-		return nil
-	}
+func reconcileCentralDBPassword(ctx context.Context, c *platform.Central, client ctrlClient.Client, statusUpdater func(updateStatusFunc), log logr.Logger) error {
 	run := &reconcileCentralDBPasswordExtensionRun{
 		SecretReconciliator: commonExtensions.NewSecretReconciliator(client, c),
-		obj:                 c,
+		centralObj:          c,
 	}
 	return run.Execute(ctx)
 }
 
 type reconcileCentralDBPasswordExtensionRun struct {
 	*commonExtensions.SecretReconciliator
-	obj *platform.Central
+	centralObj *platform.Central
+	password   string
+}
+
+func (r *reconcileCentralDBPasswordExtensionRun) readPasswordFromReferencedSecret(ctx context.Context) error {
+	if r.centralObj.Spec.Central.DB.GetPasswordSecret() == nil {
+		return errors.New("no password secret was specified")
+	}
+
+	passwordSecretName := r.centralObj.Spec.Central.DB.PasswordSecret.Name
+
+	passwordSecret := &coreV1.Secret{}
+	key := ctrlClient.ObjectKey{Namespace: r.centralObj.GetNamespace(), Name: passwordSecretName}
+	if err := r.Client().Get(ctx, key, passwordSecret); err != nil {
+		return errors.Wrapf(err, "failed to retrieve central db password secret %q", passwordSecretName)
+	}
+
+	password, err := passwordFromSecretData(passwordSecret.Data)
+	if err != nil {
+		return errors.Wrapf(err, "reading central db password from secret %s", passwordSecretName)
+	}
+
+	r.password = password
```

Comment:
Why not returning the password and setting it from the main function? It feels odd to me because the result of that function is not returned, instead in sets state on the struct.
Setting it from the caller makes it more explicit to me.